### PR TITLE
Support configure the app with generic host

### PR DESCRIPTION
**Pull request by scott-xu**
This PR should fix https://github.com/natemcmaster/CommandLineUtils/issues/309
and fix https://github.com/natemcmaster/CommandLineUtils/issues/358

### CODE REVIEW

File: src/Hosting.CommandLine/HostBuilderExtensions.cs [hunk 1]
```diff
@@ -28,12 +28,14 @@ public static class HostBuilderExtensions
         /// <typeparam name="TApp">The type of the command line application implementation</typeparam>
         /// <param name="hostBuilder">This instance</param>
         /// <param name="args">The command line arguments</param>
+        /// <param name="configure">The delegate to configure the application</param>
         /// <param name="cancellationToken">A cancellation token</param>
         /// <returns>A task whose result is the exit code of the application</returns>
         public static async Task<int> RunCommandLineApplicationAsync<TApp>(
-            this IHostBuilder hostBuilder, string[] args, CancellationToken cancellationToken = default)
+            this IHostBuilder hostBuilder, string[] args, Action<CommandLineApplication> configure = null, CancellationToken cancellationToken = default)
```
Comment (poke) [inline src/Hosting.CommandLine/HostBuilderExtensions.cs:35]: This probably needs a separate overload so that this isn’t a breaking change.

Comment (scott-xu) [reply src/Hosting.CommandLine/HostBuilderExtensions.cs]: make sense

**Comment by scott-xu** [comment]
Any objection pls?

File: src/Hosting.CommandLine/Internal/CommandLineService{T}.cs [hunk 2]
```diff
@@ -26,8 +26,9 @@ internal class CommandLineService<T> : IDisposable, ICommandLineService where T
         /// <param name="logger">A logger</param>
         /// <param name="state">The command line state</param>
         /// <param name="serviceProvider">The DI service provider</param>
+        /// <param name="configure">The delegate to configure the app</param>
         public CommandLineService(ILogger<CommandLineService<T>> logger, CommandLineState state,
-            IServiceProvider serviceProvider)
+            IServiceProvider serviceProvider, Action<CommandLineApplication> configure)
```
Comment (natemcmaster) [inline src/Hosting.CommandLine/Internal/CommandLineService{T}.cs:31]:
```suggestion
            IServiceProvider serviceProvider, Action<CommandLineApplication<T>> configure)
```

File: src/Hosting.CommandLine/HostBuilderExtensions.cs [hunk 3]
```diff
@@ -31,9 +31,33 @@ public static class HostBuilderExtensions
...
+        /// <typeparam name="TApp">The type of the command line application implementation</typeparam>
+        /// <param name="hostBuilder">This instance</param>
+        /// <param name="args">The command line arguments</param>
+        /// <param name="configure">The delegate to configure the application</param>
+        /// <param name="cancellationToken">A cancellation token</param>
+        /// <returns>A task whose result is the exit code of the application</returns>
+        public static async Task<int> RunCommandLineApplicationAsync<TApp>(
+            this IHostBuilder hostBuilder,
+            string[] args,
+            Action<CommandLineApplication> configure,
```
Comment (natemcmaster) [inline src/Hosting.CommandLine/HostBuilderExtensions.cs:56]:
```suggestion
            Action<CommandLineApplication<TApp>> configure,
```
If we make this use the generic parameter, then users can also access the `Model` and `ModelFactory` properties.

File: src/Hosting.CommandLine/Internal/CommandLineService{T}.cs (hunk 2 above)
Comment (scott-xu) [reply src/Hosting.CommandLine/Internal/CommandLineService{T}.cs:31]: The problem is the `_application` is `CommandLineApplication` but not `CommandLineApplication<T>`

Comment (natemcmaster) [reply src/Hosting.CommandLine/Internal/CommandLineService{T}.cs:31]: Is there a reason we cannot update _application to the generic type?

Comment (scott-xu) [reply src/Hosting.CommandLine/Internal/CommandLineService{T}.cs:31]: It looks very strange to me.
`_application = new CommandLineApplication<T>(state.Console, state.WorkingDirectory)`
I thought the _application is a generic one but Visual Studio tells me it is not.

Comment (scott-xu) [reply src/Hosting.CommandLine/Internal/CommandLineService{T}.cs:31]: Never mind. I forgot the variable declaration.

**Pull request merged**